File: 2017/05/rs/src/main.rs
fn do_jumps<T>(jumps: &Vec<i32>, new_jump_junc: T) -> i32
where
    T: Fn(i32) -> i32,
{
    let mut jumps = jumps.clone();
    let max_index = jumps.len() as i32;
    let mut index = 0i32;
    let mut count = 0;
    while index >= 0 && index < max_index {
        count += 1;
        let offset = jumps[index as usize];
        let next_index = index + offset;
        jumps[index as usize] = offset + new_jump_junc(offset);
        index = next_index;
    }
    count
}

fn solve(jumps: &Vec<i32>) -> (i32, i32) {
    (
        do_jumps(jumps, |_| 1),
        do_jumps(jumps, |offset| if offset < 3 { 1 } else { -1 }),
    )
}

fn get_input(file_path: &String) -> Vec<i32> {
    std::fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .lines()
        .map(|line| line.parse().unwrap())
        .collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", now.elapsed().as_secs_f32());
}

File: 2017/05/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>

typedef struct {
    int *jumps, count, size;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int part1(int offset)
{
    return 1;
}

int part2(int offset)
{
    return offset < 3 ? 1 : -1;
}

int doJumps(Input input, int (*newJumpFunc)(int))
{
    int *jumps = calloc(input.count, sizeof(int));
    memcpy(jumps, input.jumps, input.count * (sizeof(int)));
    int index = 0, count = 0, offset, nextIndex;
    while (index >= 0 && index < input.count)
    {
        count++;
        offset = jumps[index];
        nextIndex = index + offset;
        jumps[index] = offset + newJumpFunc(offset);
        index = nextIndex;
    }
    free(jumps);
    return count;
}

Results solve(Input input)
{
    return (Results){doJumps(input, &part1), doJumps(input, &part2)};
}

#define INPUT_INCREMENT 10
void addToInput(Input *input, int jump)
{
    if (input->count == input->size)
    {
        input->size += INPUT_INCREMENT;
        input->jumps = realloc(input->jumps, input->size * sizeof(int));
    }
    input->jumps[input->count++] = jump;
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    Input input = {
        calloc(INPUT_INCREMENT, sizeof(int)),
        0,
        INPUT_INCREMENT
    };
    char *line = NULL;
    size_t lineLength;
    while (getline(&line, &lineLength, file) != EOF)
        addToInput(&input, atoi(line));
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    free(input.jumps);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
